### PR TITLE
380 abstract vpc and subnet

### PR DESCRIPTION
**Pull request by adisos**
issue #380

### CODE REVIEW

**Review by zivnevo — APPROVED** [review]
I only have these nits. Feel free to ignore

File: pkg/vpcmodel/subnetsConnectivity.go [hunk 1]
```diff
@@ -129,33 +126,20 @@ func (c *VPCConfig) convertIPbasedToSubnetBasedResult(ipconn *IPbasedConnectivit
...
-		return nil, err
-	}
-	err = errors.New("unexpected res for IPblockToNamedResourcesInConfig on input subnet cidr")
-	if len(elems) != 1 {
-		return nil, err
-	}
-	if nodeSetElem, ok := elems[0].(NodeSet); ok {
-		if nodeSetElem.Kind() != subnetKind {
-			return nil, err
+func (c *VPCConfig) SubnetCidrToSubnetElem(cidr string) (Subnet, error) {
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/subnetsConnectivity.go]: moved

File: pkg/vpcmodel/grouping.go [hunk 2]
```diff
@@ -222,20 +222,20 @@ func (g *groupingConnections) addPublicConnectivity(ep EndpointElem, commonProps
 // vsiGroupingBySubnets returns a slice of EndpointElem objects, by grouping set of elements that
 // represent network interface nodes from the same subnet into a single groupedNetworkInterfaces object
 func vsiGroupingBySubnets(groupedConnLines *GroupConnLines,
-	elemsList []EndpointElem, c *VPCConfig) []EndpointElem {
+	elemsList []EndpointElem) []EndpointElem {
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/grouping.go]: changed

File: pkg/ibmvpc/vpc.go [hunk 3]
```diff
@@ -83,16 +80,20 @@ func (n *IKSNode) Name() string {
 ///////////////////////////////////////////////////////////////////////////////////////////////////
 // nodesets elements - implement vpcmodel.NodeSet interface
 
+// VPC implements vpcmodel.VPC
 type VPC struct {
 	vpcmodel.VPCResource
 	nodes                []vpcmodel.Node
-	connectivityRules    *vpcmodel.ConnectivityResult // allowed connectivity between elements within the vpc
```
Comment (ShiriMoran) [inline pkg/ibmvpc/vpc.go:89]: Why was this removed? What replaces it?

File: pkg/ibmvpc/vpc.go [hunk 4]
```diff
@@ -115,12 +113,16 @@ func (v *VPC) subnets() []*Subnet {
 	return v.subnetsList
 }
 
+// Subnet implements vpcmodel.Subnet interface
 type Subnet struct {
 	vpcmodel.VPCResource
-	nodes             []vpcmodel.Node
-	connectivityRules *vpcmodel.ConnectivityResult // allowed connectivity between elements within the subnet
```
Comment (ShiriMoran) [inline pkg/ibmvpc/vpc.go:121]: Same question - why was this removed? what replaces it?

File: pkg/vpcmodel/abstractVPC.go [hunk 5]
```diff
@@ -130,14 +138,23 @@ func (n *InternalNode) IsPublicInternet() bool {
...
 }
 
+type VPC interface {
+	NodeSet
+	AddressPrefixes() []string
+}
+
+type Subnet interface {
+	NodeSet
+	CIDR() string
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/abstractVPC.go:155]: I already embeds `NodeSet` which has method `AddressRange() *ipblocks.IPBlock`

**Review by zivnevo — APPROVED** [review]

File: pkg/ibmvpc/parser.go [hunk 6]
```diff
@@ -908,12 +902,8 @@ func getIKSnodesConfig(res map[string]*vpcmodel.VPCConfig,
 
 func NewEmptyVPCConfig() *vpcmodel.VPCConfig {
 	return &vpcmodel.VPCConfig{
-		Nodes:            []vpcmodel.Node{},
-		NodeSets:         []vpcmodel.NodeSet{},
-		FilterResources:  []vpcmodel.FilterTrafficResource{},
-		RoutingResources: []vpcmodel.RoutingResource{},
-		UIDToResource:    map[string]vpcmodel.VPCResourceIntf{},
-		CloudName:        "IBM Cloud",
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/parser.go:916]: only the slices initialization was removed, those are not required (can append to nil slice).

File: pkg/ibmvpc/vpc.go (hunk 3 above)
Comment (adisos) [reply pkg/ibmvpc/vpc.go:89]: this field was not in use

File: pkg/ibmvpc/vpc.go (hunk 4 above)
Comment (adisos) [reply pkg/ibmvpc/vpc.go:121]: same here

File: pkg/vpcmodel/vpcConfig.go [hunk 7]
```diff
@@ -8,9 +8,14 @@ import (
...
 type VPCConfig struct {
+	// VPC is a reference to the relevant VPC object for which this config belongs
+	VPC VPC
+	// Subnets is a list of subnets in the VPC
+	Subnets []Subnet
 	// Nodes is the list of endpoints in the VPC, such as network interfaces, reserved IPs
+	// TOTO: also separate Nodes to internal and external nodes lists
 	Nodes []Node
-	// NodeSets is the list of resources that capture multiple nodes, such as subnets, vsis, vpc
+	// NodeSets is the list of resources that capture multiple nodes (such as VSIs)
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/vpcConfig.go]: updated

File: pkg/vpcmodel/abstractVPC.go [hunk 8]
```diff
@@ -130,14 +138,23 @@ func (n *InternalNode) IsPublicInternet() bool {
...
 type NodeSet interface {
 	VPCResourceIntf
 	Nodes() []Node
-	Connectivity() *ConnectivityResult
 	AddressRange() *ipblocks.IPBlock
 }
 
+type VPC interface {
+	NodeSet
+	AddressPrefixes() []string
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/abstractVPC.go:150]: Also, note that now you can use this method for explainability input.

File: pkg/vpcmodel/vpcConfig.go [hunk 9]
```diff
@@ -8,9 +8,14 @@ import (
 
 // VPCConfig captures the configured resources for a VPC
 type VPCConfig struct {
+	// VPC is a reference to the relevant VPC object for which this config belongs
+	VPC VPC
+	// Subnets is a list of subnets in the VPC
+	Subnets []Subnet
 	// Nodes is the list of endpoints in the VPC, such as network interfaces, reserved IPs
+	// TOTO: also separate Nodes to internal and external nodes lists
```
Comment (ShiriMoran) [inline pkg/vpcmodel/vpcConfig.go:18]:
```suggestion
	// TODO: also separate Nodes to internal and external nodes lists
```

File: pkg/ibmvpc/vpc.go [hunk 10]
```diff
@@ -361,66 +350,85 @@ func (n *NACL) GeneralConnectivityPerSubnet(subnet *Subnet) string {
...
 }
 
+type naclConnectivityInput struct {
+	targetNode           vpcmodel.Node
+	nodeInSubnet         vpcmodel.Node
+	subnet               *Subnet
+	subnetAffectedByNACL bool
+	targetWithinSubnet   bool
+}
+
```
Comment (ShiriMoran) [inline pkg/ibmvpc/vpc.go:367]: This is yet another optimization not related to the specific abstraction added in this PR. right?

File: pkg/ibmvpc/vpc.go [hunk 11]
```diff
@@ -361,66 +350,85 @@ func (n *NACL) GeneralConnectivityPerSubnet(subnet *Subnet) string {
...
 func (n *NACL) initConnectivityComputation(src, dst vpcmodel.Node,
-	isIngress bool) (targetNode, nodeInSubnet vpcmodel.Node, subnet *Subnet, err error) {
+	isIngress bool) (
```
Comment (ShiriMoran) [inline pkg/ibmvpc/vpc.go:369]: Also here, optimization which is not related to the  new abstraction. Right? or am I missing something

File: pkg/ibmvpc/vpc.go (hunk 10 above)
Comment (adisos) [reply pkg/ibmvpc/vpc.go:367]: right

File: pkg/ibmvpc/vpc.go (hunk 11 above)
Comment (adisos) [reply pkg/ibmvpc/vpc.go:369]: right

File: pkg/ibmvpc/vpc.go (hunk 10 above)
Comment (ShiriMoran) [reply pkg/ibmvpc/vpc.go:367]: next time it would be nice to have it in a different PR :-)

**Review by ShiriMoran — APPROVED** [review]
